### PR TITLE
optionally persist metadata sends from metadata plugin

### PR DESCRIPTION
**Pull request by codesmith14**
* save/load metadata send rate state to a json file

### CODE REVIEW

**Review by jrcamp — APPROVED** [review]

File: src/signalfx_metadata.py [hunk 1]
```diff
@@ -97,6 +97,8 @@
 PERCORECPUUTIL = False
 OVERALLCPUUTIL = True
 ETC_PATH = "{0}etc".format(os.sep)
+PERSISTENCE_PATH = None
+PERSISTENCE_FILE = "persist_sfx_metadata_state.js"
```
Comment (mdubbyap) [inline src/signalfx_metadata.py]: .json?

File: src/signalfx_metadata.py [hunk 2]
```diff
@@ -883,6 +890,39 @@ def plugin_config(conf):
...
+        with open(os.path.join(PERSISTENCE_PATH, PERSISTENCE_FILE), 'r') as js:
+            persist = json.load(js)
+            debug("Loaded the following persistent data %s" % persist)
+            if "NEXT_METADATA_SEND" in persist:
+                global NEXT_METADATA_SEND
+                NEXT_METADATA_SEND = persist["NEXT_METADATA_SEND"]
+            if "NEXT_METADATA_SEND_INTERVAL" in persist:
+                global NEXT_METADATA_SEND_INTERVAL
+                NEXT_METADATA_SEND_INTERVAL = \
+                    persist["NEXT_METADATA_SEND_INTERVAL"]
```
Comment (mdubbyap) [inline src/signalfx_metadata.py]: you need to store and check the hostname against what this running one thinks is the hostname. if it's different you need to restart the metadata send intervals.

**Review by mdubbyap — CHANGES_REQUESTED** [review]

File: src/signalfx_metadata.py (hunk 2 above)
Comment (codesmith14) [reply src/signalfx_metadata.py]: @mdubbyap I have addressed your comment regarding host name.  Please see the latest commit to this pr.  Host name is now loaded and saved, and then checked when `steal_host_from_notifications()` is invoked.  If the loaded host name doesn't match, the metadata interval and next metadata send time are reset.  This should reset things prior to the first time the metadata code path is executed.

File: src/signalfx_metadata.py [hunk 3]
```diff
@@ -883,6 +905,43 @@ def plugin_config(conf):
     get_aws_info()
 
 
+def load_persistent_data():
+    """Load persistent data from a specified path"""
+    try:
+        with open(os.path.join(PERSISTENCE_PATH, PERSISTENCE_FILE), 'r') as js:
+            persist = json.load(js)
+            debug("Loaded the following persistent data %s" % persist)
+            if "SAVED_HOST" in persist:
```
Comment (mdubbyap) [inline src/signalfx_metadata.py]: all these strings used multiple times should be constants

File: src/signalfx_metadata.py [hunk 4]
```diff
@@ -883,6 +905,43 @@ def plugin_config(conf):
...
+def save_persistent_data():
+    """Persist data about the metadata plugin to a file"""
+    if PERSISTENCE_PATH:
+        try:
+            with open(os.path.join(PERSISTENCE_PATH,
+                                   PERSISTENCE_FILE), 'w') as f:
+                persist = {
+                    'SAVED_HOST': HOST,
+                    'NEXT_METADATA_SEND': NEXT_METADATA_SEND,
+                    'NEXT_METADATA_SEND_INTERVAL': NEXT_METADATA_SEND_INTERVAL
```
Comment (mdubbyap) [inline src/signalfx_metadata.py]: more places where the constants should be used

**Review by mdubbyap — CHANGES_REQUESTED** [review]
Your changes look good.  One small thing before i'm satisfied.

**Review by mdubbyap — APPROVED** [review]
Squash your commits and merge. Thanks.

**Pull request merged**